infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServerNode.java

```diff
@@ -136,30 +139,36 @@ public void close() {
 
         // A executor service to create the shutdown threads
         // plus name the threads correctly.
-        final ExecutorService shutdownService = Executors.newFixedThreadPool(serverMap.size(),
-                new ServerThreadFactory("CorfuServer-shutdown-",
-                        new ServerThreadFactory.ExceptionHandler()));
+        final ExecutorService shutdownService = Executors.newFixedThreadPool(
+                serverMap.size(),
+                new ServerThreadFactory("CorfuServer-shutdown-", new ExceptionHandler())
+        );
 
         // Turn into a list of futures on the shutdown, returning
         // generating a log message to inform of the result.
-        CompletableFuture[] shutdownFutures = serverMap.values().stream()
-                .map(server -> CompletableFuture.runAsync(() -> {
-                    try {
-                        log.info("close: Shutting down {}", server.getClass().getSimpleName());
-                        server.shutdown();
-                        log.info("close: Cleanly shutdown {}", server.getClass().getSimpleName());
-                    } catch (Exception e) {
-                        log.error("close: Failed to cleanly shutdown {}",
-                                server.getClass().getSimpleName(), e);
-                    }
-                }, shutdownService))
-                .toArray(CompletableFuture[]::new);
+        List<CompletableFuture<Void>> shutdownFutures = serverMap.values()
+                .stream()
+                .map(server -> shutdownServer(shutdownService, server))
+                .collect(Collectors.toList());
 
-        CompletableFuture.allOf(shutdownFutures).join();
+        CFUtils.allOf(shutdownFutures).join();
```

Comment:
I would suggest we time out and log a warning if its not done within a some period of time. Indefinitely waiting during shutdown seems suspicious.

Reply:
Not in this bug fix

Reply:
fixed

Reply:
We are still waiting on this infinitely, but all the servers now fail by default timeout. Please take a look